File: code/R/proportion_with_unsuitable_underlying.R
# Load libraries and supporting functions ----

# Load necessary libraries
library(here)

# Load supporting functions
source(here::here("code", "R", "dqaf_metrics.R"))

# Specify data and column names ----

# Load the death records data
file_path <- here::here("data", "SyntheticDeathRecordData.csv")
death_records <- 
  load_death_records(file_path)

# Specify required metric column names here
underlying_cause_of_death_column <- "Underlying COD"

# Calculate metric ----

# Load the unsuitable causes of death data
unsuitable_causes <- read.csv(here::here("data", "unsuitable_COD_codes.csv"))

# Extract the unsuitable codes
unsuitable_codes <- unsuitable_causes$code

# Create a new column that is TRUE when the underlying COD is unsuitable
death_records[, "Unsuitable Underlying"] <- sapply(
  death_records[, underlying_cause_of_death_column], 
  function(code){
    return(any(startsWith(code, unsuitable_codes)))
  }
)

# Calculate the proportion of records with an unsuitable underlying cause of death
proportion <- calculate_proportion(
  death_records, 
  metric = "Unsuitable Underlying",
  metric_description = "unsuitable underlying cause of death", 
  print_output = TRUE
)

# Group the records by certifier and calculate the proportion of unsuitable records for each certifier
certifier_proportions <- calculate_proportion_by_column(
  death_records, 
  metric = "Unsuitable Underlying",
  column = "Certifier Name", 
  print_output = TRUE
)

File: code/python/proportion_with_unsuitable_underlying.py
import os
import pandas as pd

# Locate the relative path to the data location
data_path = os.path.join(os.path.dirname(os.path.realpath(__file__)), "..", "..", "data")

# Load the death records data
death_records = pd.read_csv(
    os.path.join(data_path, "SyntheticDeathRecordData.csv"),
    keep_default_na=False,
    na_values=[""],
)

# Load the unsuitable causes of death data
unsuitable_causes = pd.read_csv(os.path.join(data_path, "unsuitable_COD_codes.csv"))

# Extract the unsuitable codes
unsuitable_codes = unsuitable_causes["code"].values


# Function to check if any unsuitable code is a prefix to the code in the record
def is_unsuitable(code):
    return any(code.startswith(unsuitable) for unsuitable in unsuitable_codes)


# Create a new column that is True when the underlying COD is unsuitable
death_records["Unsuitable Underlying"] = death_records["Underlying COD"].apply(
    is_unsuitable
)

# Calculate the proportion of records with an unsuitable underlying cause of death
proportion = death_records["Unsuitable Underlying"].mean()

print(
    f"The proportion of records with an unsuitable underlying cause of death is {proportion:.2f}"
)

# Group the records by certifier and calculate the proportion of unsuitable records for each certifier
certifier_proportions = death_records.groupby("Certifier Name")[
    "Unsuitable Underlying"
].mean()

# Print the proportions for each certifier
for certifier, proportion in certifier_proportions.items():
    print(
        f"The proportion of records with an unsuitable underlying cause of death provided by {certifier} is {proportion:.2f}"
    )
